api/authentication.py

```diff
@@ -1,160 +1,414 @@
-        raise AuthenticationFailed("FXA returned active: False for token.")
-
-    # FxA user is active, check for the associated Relay account
-    if (raw_fxa_uid := fxa_resp_data.get("json", {}).get("sub")) is None:
-        raise NotFound("FXA did not return an FXA UID.")
-    fxa_uid = str(raw_fxa_uid)
-
-    # cache valid access_token and fxa_resp_data until access_token expiration
-    # TODO: revisit this since the token can expire before its time
-    if isinstance(fxa_resp_data.get("json", {}).get("exp"), int):
-        # Note: FXA iat and exp are timestamps in *milliseconds*
-        fxa_token_exp_time = int(fxa_resp_data["json"]["exp"] / 1000)
-        now_time = int(datetime.now(UTC).timestamp())
-        fxa_token_exp_cache_timeout = fxa_token_exp_time - now_time
-        if fxa_token_exp_cache_timeout > cache_timeout:
-            # cache until access_token expires (matched Relay user)
-            # this handles cases where the token already expired
-            cache_timeout = fxa_token_exp_cache_timeout
-    cache.set(cache_key, fxa_resp_data, cache_timeout)
-
-    return fxa_uid
-
-
-class FxaTokenAuthentication(BaseAuthentication):
-    def authenticate_header(self, request):
-        # Note: we need to implement this function to make DRF return a 401 status code
-        # when we raise AuthenticationFailed, rather than a 403. See:
-        # https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
-        return "Bearer"
-
-    def authenticate(self, request):
-        authorization = get_authorization_header(request).decode()
-        if not authorization or not authorization.startswith("Bearer "):
-            # If the request has no Bearer token, return None to attempt the next
-            # auth scheme in the REST_FRAMEWORK AUTHENTICATION_CLASSES list
-            return None
-
-        token = authorization.split(" ")[1]
-        if token == "":
-            raise ParseError("Missing FXA Token after 'Bearer'.")
-
-        use_cache = True
+    if not isinstance(sub := data.get("sub", None), str) or not sub:
+        return IntrospectionError("NoSubject", status_code=status_code, data=fxa_data)
+
+    return IntrospectionResponse(data=cast(FxaIntrospectData, data))
+
+
+def introspect_token_or_raise(
+    token: str, use_cache: bool = True
+) -> IntrospectionResponse:
+    """
+    Introspect a Mozilla account OAuth token, to get data like the FxA UID.
+
+    If anything goes wrong, raise an exception.
+    """
+    default_cache_timeout = 60
+
+    fxa_resp: IntrospectionResponse | IntrospectionError | None = None
+    if use_cache:
+        fxa_resp = load_introspection_result_from_cache(cache, token)
+    if fxa_resp is None:
+        fxa_resp = introspect_token(token)
+
+    # If the response is an error, raise an exception
+    if isinstance(fxa_resp, IntrospectionError):
+        if not fxa_resp.from_cache:
+            fxa_resp.save_to_cache(cache, token, default_cache_timeout)
```

Comment:
question (non-blocking): If I understand correctly, this means we will cache introspection errors for just 60s? I like that. When I first saw the cache + error code, I thought it might cache something like a timeout error for a long time. This looks like it will prevent a thundering herd AND recover from temporary errors in a reasonable amount of time.